### PR TITLE
net: lib: nrf_cloud: use new c2d topics

### PR DESCRIPTION
**Pull request by jayteemo**
Use new topics to receive data from nRF Cloud.
Data for location services now comes in on separate topics to allow for better filtering.
A-GPS and P-GPS is no longer forwarded to the application. Cellular positioning data is forwarded only if a callback has not been registered.
Created new event IDs for RX data.
NCSDK-11891

Signed-off-by: Justin Morton <justin.morton@nordicsemi.no>

### CODE REVIEW

**Comment by jayteemo** [comment]
draft until cloud changes are live on prod.
testable on dev.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c
Comment (plskeggs) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:590]: It would be nice to have the topic logged here, ideally with the data, though that's problematic for A-GPS.

**Comment by plskeggs** [comment]
Looks good so far.  Certainly simplifies the integration of A-GPS and P-GPS.

**Review by glarsennordic — COMMENTED** [review]
I like it so far too!

**Comment by jayteemo** [comment]
Note: this PR is "complete", i'm not making any changes other than rebasing and addressing PR comments.
it is only in draft because the cloud changes are not on prod.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c
Comment (jayteemo) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:897]: Added [requested topic logging](https://github.com/nrfconnect/sdk-nrf/pull/8780#discussion_r974716168) here @plskeggs

File: applications/asset_tracker_v2/src/modules/cloud_module.c
Comment (glarsennordic) [inline applications/asset_tracker_v2/src/modules/cloud_module.c:292]: Debug logs for these could be useful

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c
Comment (glarsennordic) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:696]: Nit: Most of the logs in this file are capitalized

Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:696]: Consider capitalizing the "could not decode" occurance on 671 as well

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c
Comment (glarsennordic) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:553]: Remind me again why this must be called in response to agps data receipt? I don't question it, but I would like to fully understand

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c
Comment (glarsennordic) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:574]: It seems like the contents of `cell_pos_cb_send` should just be moved into this function

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c
Comment (glarsennordic) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:615]: I feel that this should have its own topic (a device control topic), but given that that would require further BE changes, maybe that would be for a round-two PR

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c
Comment (glarsennordic) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:900]: This will be nice to have, though I'm not sure whether name is guaranteed to be null-terminated. You may want to use `%.*s` with `message.topic.topic.size`. Or do you know for sure it is null-terminated?

**Comment by jayteemo** [comment]
added a commit for error events based on a comment from @simensrostad

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:571]: i don't know, but i bet @plskeggs does

Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:571]: If both A-GPS and P-GPS are enabled, everything but ephemerides and almanacs are handled by A-GPS.  In this configuration, it is important to check, after receiving A-GPS data, whether any further assistance is needed by the modem for ephemerides, which would come from P-GPS (usually, in stored predictions in flash).

Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:571]: maybe that should be a comment in the code somewhere?  or maybe it already is?

File: applications/asset_tracker_v2/src/modules/cloud_module.c
Comment (jayteemo) [reply applications/asset_tracker_v2/src/modules/cloud_module.c:292]: code was just moved around for this change.  i'll leave that up to @simensrostad in a future atv2 update.

Comment (glarsennordic) [reply applications/asset_tracker_v2/src/modules/cloud_module.c:292]: Ah you're right. I lost sight of the fact I was looking at ATV2 code not library code.

**Review by glarsennordic — APPROVED** [review]
LGTM

**Review by plskeggs — APPROVED** [review]

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c
Comment (junqingzou) [inline applications/serial_lte_modem/src/gnss/slm_at_gnss.c:769]: Need below condition be satisfied

if (nrf_cloud_ready) {

**Comment by junqingzou** [comment]
Is it correct understanding that APP is still responsible to make A-GPS and P-GPS request to nRF Cloud (by calling lib_nrf_cloud API), but the response is now all handled in side the lib_nrf_cloud? Will lib_nrf_cloud send back an event with the result of handling A-GPS and P-GPS request?

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c
Comment (jayteemo) [reply applications/serial_lte_modem/src/gnss/slm_at_gnss.c:769]: i don't think that is needed anymore, but i will add it.
this is the handler for the new event `NRF_CLOUD_EVT_RX_DATA_GENERAL`; there should not be any of these events before the cloud is ready.
a new event `NRF_CLOUD_EVT_RX_DATA_SHADOW` may occur before the cloud is ready, but this data is ignored by slm.

**Comment by jayteemo** [comment]
> Will lib_nrf_cloud send back an event with the result of handling A-GPS and P-GPS request?

Only on a processing error.
Is there a use case for the application to receive an event indicating that A/P-GPS data has been processed?

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c
Comment (junqingzou) [reply applications/serial_lte_modem/src/gnss/slm_at_gnss.c:769]: Yes it was the shadow data that confused customers before. Logically we should only allow uplink/downlink "general" after `NRF_CLOUD_EVT_READY`, so please add the `if` check even it's redundant now.

**Comment by junqingzou** [comment]
> > Will lib_nrf_cloud send back an event with the result of handling A-GPS and P-GPS request?
> 
> Only on a processing error. Is there a use case for the application to receive an event indicating that A/P-GPS data has been processed?

At least currently not required for SLM application, just for the sake of request-response model.

File: include/net/nrf_cloud.h
Comment (jayteemo) [reply include/net/nrf_cloud.h:140]: i was copying the general format of the source error enum:
https://github.com/nrfconnect/sdk-zephyr/blob/97c4a1a6d1bda83f0272d9bbf68b211efadc7b4a/include/zephyr/net/mqtt.h#L130

**Review by junqingzou — APPROVED** [review]
Changes in SLM/GNSS are good

**Review by simensrostad — APPROVED** [review]

**Comment by jayteemo** [comment]
Removing draft status. The required cloud BE changes to prod are about a week away.  Labeling with `DNM` until then.

**Comment by trantanen** [comment]
When prod cloud is updated to support these changes, will it continue to handle older requests/responses from older implementations?

File: applications/asset_tracker_v2/src/cloud/nrf_cloud_integration.c
Comment (trantanen) [inline applications/asset_tracker_v2/src/cloud/nrf_cloud_integration.c:237]: I have another PR where I'm reading cell pos results in ATv2. I guess I can find them still at `evt->data.ptr`, right?

File: subsys/net/lib/nrf_cloud/include/nrf_cloud_codec.h
Comment (tokangas) [reply subsys/net/lib/nrf_cloud/include/nrf_cloud_codec.h:137]: UNKNOWN sounds good to me.

**Review by tokangas — APPROVED** [review]
Looks good.

File: applications/asset_tracker_v2/src/cloud/nrf_cloud_integration.c
Comment (jayteemo) [reply applications/asset_tracker_v2/src/cloud/nrf_cloud_integration.c:237]: correct

**Comment by jayteemo** [comment]
> When prod cloud is updated to support these changes, will it continue to handle older requests/responses from older implementations?

correct

File: doc/nrf/releases/release-notes-changelog.rst
Comment (peknis) [inline doc/nrf/releases/release-notes-changelog.rst:378]:
```suggestion
    * The status field of :c:enum:`NRF_CLOUD_EVT_ERROR` events use values from the enumeration :c:enumerator:`nrf_cloud_error_status`.
```

**Review by trantanen — APPROVED** [review]

File: include/net/nrf_cloud.h
Comment (peknis) [inline include/net/nrf_cloud.h:181]:
```suggestion
 * See the <a href="https://api.nrfcloud.com/v1#section/Error-Codes">Error Codes</a> section of nRF Cloud API documentation for
```

File: doc/nrf/releases/release-notes-changelog.rst
Comment (peknis) [inline doc/nrf/releases/release-notes-changelog.rst:376]:
```suggestion
     * Added events :c:enum:`NRF_CLOUD_EVT_RX_DATA_CELL_POS` and :c:enum:`NRF_CLOUD_EVT_RX_DATA_SHADOW`.
```

**Review by peknis — CHANGES_REQUESTED** [review]

**Review by peknis — APPROVED** [review]

**Comment by trantanen** [comment]
> Removing draft status. The required cloud BE changes to prod are about a week away. Labeling with `DNM` until then.

What is the expected timeline for this PR to be merged? That is, related cloud changes to be in prod?

**Review by jhirsi — APPROVED** [review]

File: subsys/net/lib/nrf_cloud/include/nrf_cloud_codec.h
Comment (peknis) [inline subsys/net/lib/nrf_cloud/include/nrf_cloud_codec.h:247]: A general note on the verb form in all function descriptions. You should always use the imperative format: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/doc_styleguide.html#functions

This header file seems to have most of them in the third person format.

**Review by peknis — COMMENTED** [review]
Left a general note on the function descriptions.

File: doc/nrf/releases/release-notes-changelog.rst
Comment (melwee01) [inline doc/nrf/releases/release-notes-changelog.rst:475]: After some discussion, we decided the abbreviations for C2D, D2C, and so on should be with capital letters.

File: doc/nrf/releases/release-notes-changelog.rst
Comment (melwee01) [inline doc/nrf/releases/release-notes-changelog.rst:476]:
```suggestion
      This enables the device to receive Location Services data on separate topics.
```

Generically referring to 'location data' would also be okay, but Location Services is a capitalized product name.

File: doc/nrf/releases/release-notes-changelog.rst
Comment (melwee01) [inline doc/nrf/releases/release-notes-changelog.rst:480]:
```suggestion
    * The status field of :c:enum:`NRF_CLOUD_EVT_ERROR` events uses values from the enumeration :c:enumerator:`nrf_cloud_error_status`.
```

One field, 'of events' doesn't change verb number.

File: samples/nrf9160/nrf_cloud_mqtt_multi_service/README.rst
Comment (melwee01) [inline samples/nrf9160/nrf_cloud_mqtt_multi_service/README.rst:166]: I'd prefer 'involves' to something like 'amounts to'.

**Review by melwee01 — COMMENTED** [review]
These are small points, so I don't want to hold up merging, but the branding capitalization definitely needs fixing. 

Perhaps check that there aren't other instances in the same files where Location Services needs to be capitalized.

File: doc/nrf/releases/release-notes-changelog.rst
Comment (jayteemo) [reply doc/nrf/releases/release-notes-changelog.rst:475]: the topic string is `"c2d"`, so I would prefer to keep it lower case.
how should i format it? just add quotes?

Comment (melwee01) [reply doc/nrf/releases/release-notes-changelog.rst:500]: No quotes or other formatting, just C2D

If you want to keep it lowercase, use ``/c2d`` to refer to the exact topic text.

**Pull request merged**
